src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs

```diff
@@ -371,6 +380,21 @@ void Complete()
             }
         }
 
+        private static bool IsAssignableToGenericTaskType(Type? t)
+        {
+            while (t != null)
```

Comment:
If I'm not mistaken, this will be done on every task completion? Are we sure it's cheap enough to do that instead of caching it per-type?

Reply:
> this will be done on every task completion?

Correct.

> Are we sure it's cheap enough to do that instead of caching it per-type?

We are currently getting the MethodInfo for the `get_Result` method on every task completion, which is a much more intensive operation than walking the Type hierarchy.

For the typical case, `t` will directly be a `Task<something>` and so the only real cost here is the method call and setting up the loop. The first time through the loop, this method will return `true`, since the Type is typically `Task<something>`.

> caching it per-type?

This would end up being a pretty big cache, as there could be all sorts of types being passed here - `Task<int>`, `Task<string>`, `Task<Foo>`, etc. If we were to implement a cache, we should be caching the `MethodInfo` for `get_Result` instead.

Reply:
It just occured to me that we could cache the result of `typeof(Task<>).GetMethod("get_Result")` in a static somewhere.

Then do the `task_type.GetMethod("get_Result")` like usual, but before we invoke the method, we would call `HasSameDefinitionAs` to check the get_Result method we got is the one we expect.

Then we don't need this base hierarchy walk, we fix the reliability issue with a random `get_Result` method getting in the way, and we don't need the DynamicDependency anymore because the cached place is easy to analyze for linker.